### ISSUE TITLE
SaveChanges causes NullReferenceException if composite Primary Key is null

### ISSUE THREAD

**Issue by Suchiman** [aspnet/EntityFramework6#559]
When you try to save an Entity that has a 3 column composite Primary Key and the property in the middle (position 2) is null, then you'll get following `NullReferenceException`.
If the third column of the Primary Key is null, you'll instead receive a helpful `DbEntityValidationException`.

```
Exception message: Object reference not set to an instance of an object
Stack trace:
   at System.Data.Entity.Core.Objects.EntityEntry.EntityMemberChanged(String entityMemberName, Object complexObject, String complexObjectMemberName)
   at System.Data.Entity.Core.Objects.EntityEntry.EntityMemberChanged(String entityMemberName)
   at System.Data.Entity.Core.Objects.ObjectStateEntry.System.Data.Entity.Core.Objects.DataClasses.IEntityChangeTracker.EntityMemberChanged(String entityMemberName)
   at System.Data.Entity.Core.Objects.Internal.SnapshotChangeTrackingStrategy.SetCurrentValue(EntityEntry entry, StateManagerMemberMetadata member, Int32 ordinal, Object target, Object value)
   at System.Data.Entity.Core.Objects.Internal.EntityWrapper`1.SetCurrentValue(EntityEntry entry, StateManagerMemberMetadata member, Int32 ordinal, Object target, Object value)
   at System.Data.Entity.Core.Objects.EntityEntry.SetCurrentEntityValue(StateManagerTypeMetadata metadata, Int32 ordinal, Object userObject, Object newValue)
   at System.Data.Entity.Core.Objects.ObjectStateEntryDbUpdatableDataRecord.SetRecordValue(Int32 ordinal, Object value)
   at System.Data.Entity.Core.Mapping.Update.Internal.PropagatorResult.SetServerGenValue(Object value)
   at System.Data.Entity.Core.Mapping.Update.Internal.UpdateTranslator.BackPropagateServerGen(List`1 generatedValues)
   at System.Data.Entity.Core.Mapping.Update.Internal.UpdateTranslator.Update()
   at System.Data.Entity.Core.EntityClient.Internal.EntityAdapter.<Update>b__2(UpdateTranslator ut)
   at System.Data.Entity.Core.EntityClient.Internal.EntityAdapter.Update[T](T noChangesResult, Func`2 updateFunction)
   at System.Data.Entity.Core.EntityClient.Internal.EntityAdapter.Update()
   at System.Data.Entity.Core.Objects.ObjectContext.<SaveChangesToStore>b__35()
   at System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChangesToStore(SaveOptions options, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction)
   at System.Data.Entity.Core.Objects.ObjectContext.<>c__DisplayClass2a.<SaveChangesInternal>b__27()
   at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute[TResult](Func`1 operation)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChangesInternal(SaveOptions options, Boolean executeInExistingTransaction)
   at System.Data.Entity.Core.Objects.ObjectContext.SaveChanges(SaveOptions options)
   at System.Data.Entity.Internal.InternalContext.SaveChanges()
   at System.Data.Entity.Internal.LazyInternalContext.SaveChanges()
   at System.Data.Entity.DbContext.SaveChanges()
   at EfTest.Program.Main(String[] args) in Program.cs: Zeile31
```

### Steps to reproduce
1. Create a new Console Application and Install EF 6.2
2. Use following Code
    ```c#
    using System.Collections.Generic;
    using System.ComponentModel.DataAnnotations;
    using System.ComponentModel.DataAnnotations.Schema;
    using System.Data.Entity;
    
    namespace EfTest
    {
        class Program
        {
            static void Main(string[] args)
            {
                var root = new Root();
                root.ItemLists.Add(new ItemList
                {
                    ListName = "SomeName",
                    Items = new List<ListItem>
                    {
                        new ListItem
                        {
                            Root = root,
                            //ListName = "SomeName",
                            Value = "SomeValue"
                        }
                    }
                });
    
                Database.SetInitializer(new MigrateDatabaseToLatestVersion<TestContext, Migrations.Configuration>());
    
                var context = new TestContext();
                context.Roots.Add(root);
                context.SaveChanges();
            }
        }
    
        public class TestContext : DbContext
        {
            public virtual DbSet<Root> Roots { get; set; }
            public virtual DbSet<ListItem> ListItems { get; set; }
            public virtual DbSet<ItemList> ItemLists { get; set; }
    
            public TestContext() : base("Server=.;Database=EfTest;Trusted_Connection=True;MultipleActiveResultSets=True")
            {
            }
    
            protected override void OnModelCreating(DbModelBuilder map)
            {
                map.Entity<ListItem>()
                    .HasRequired(x => x.ItemList)
                    .WithMany(x => x.Items)
                    .WillCascadeOnDelete(false);
            }
        }
    
        public class Root
        {
            public virtual int Id { get; set; }
            public virtual ICollection<ItemList> ItemLists { get; set; } = new List<ItemList>();
        }
    
        public class ItemList
        {
            [Key]
            [Column(Order = 1)]
            public virtual int RootId { get; set; }
            public virtual Root Root { get; set; }
    
            [Key]
            [Column(Order = 2)]
            public virtual string ListName { get; set; }
    
            public virtual ICollection<ListItem> Items { get; set; } = new List<ListItem>();
        }
    
        public class ListItem
        {
            [Key]
            [Column(Order = 1)]
            public virtual int RootId { get; set; }
            public virtual Root Root { get; set; }
    
            [Key]
            [Column(Order = 2)]
            [StringLength(50)]
            public virtual string ListName { get; set; }
            public virtual ItemList ItemList { get; set; }
    
            [Key]
            [Column(Order = 3)]
            [StringLength(90)]
            public virtual string Value { get; set; }
        }
    }
    ```
3. `Enable-Migrations`
4. `Add-Migration Initial`
5. Run the project

If you uncomment `//ListName = "SomeName",` the SaveChanges will be successfull and if you comment `Value = "SomeValue"` you'll receive the `DbEntityValidationException` instead

### Further technical details
EF version:
```csproj
<PackageReference Include="EntityFramework">
  <Version>6.2.0</Version>
</PackageReference>
```
Database Provider: EntityFramework.SqlServer
Operating system: Windows 10 1709
IDE: Visual Studio 2017 15.7.3

**Comment by ajcvickers** [comment]
@Suchiman EF treats a composite property value as null if any of its parts are null. Doing otherwise adds more complexity than is at first apparent because it means that EF has to do null comparisons on key columns on the database side, which quickly runs into ambiguities as to what matches and what doesn't.

**Comment by ajcvickers** [comment]
Sorry, I answered to quickly. I see you're just talking about the exception. In which case:

**EF Team Triage:** This issue is not something that our team is planning to address in the EF6.x code base.  This does not mean that we would not consider a community contribution to address this issue.

Moving forwards, our team will be fixing bugs, implementing small improvements, and accepting community contributions to the EF6.x code base. Larger feature work and innovation will happen in the EF Core code base (https://github.com/aspnet/EntityFramework).

**Closing an issue in the EF6.x project does not exclude us addressing it in EF Core. In fact, a number of popular feature requests for EF have already been implemented in EF Core (alternate keys, batching in SaveChanges, etc.).**

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*
